cpp/include/cuspatial/cubic_spline.hpp

```diff
@@ -0,0 +1,102 @@
+ * identify which specific spline a given query_point is interpolated with.
+ *
+ * @param[in] coefficients table of spline coefficients produced by
+ * cubicspline_coefficients.
+ *
+ * @return cudf::column `y` coordinates interpolated from `x` and `coefs`.
+**/
+std::unique_ptr<cudf::column> cubicspline_interpolate(
+                                         cudf::column_view const& query_points,
+                                         cudf::column_view const& spline_ids,
+                                         cudf::column_view const& offsets,
+                                         cudf::column_view const& source_points,
+                                         cudf::table_view const& coefficients,
+                                         rmm::mr::device_memory_resource *mr,
+                                         cudaStream_t stream);
+
+std::unique_ptr<cudf::column> cubicspline_interpolate_default(
+                                         cudf::column_view const& query_points,
+                                         cudf::column_view const& spline_ids,
+                                         cudf::column_view const& offsets,
+                                         cudf::column_view const& source_points,
+                                         cudf::table_view const& coefficients);
+
+/**
+ * @brief Create a table of cubic spline coefficients from columns of coordinates.
+ *
+ * Computes coefficients for a natural cubic spline similar to the method
+ * found on http://mathworld.wolfram.com/CubicSpline.html .
+ *
+ * The input data arrays `t` and `y` contain the vertices of many concatenated
+ * splines.
+ *
+ * Currently, all input splines must be the same length. The minimum supported
+ * length is 5.
+ * 
+ * @note Ids should be prefixed with a 0, even when only a single spline
+ * is fit, ids will be {0, 0}
+ *
+ * @param[in] t column_view of independent coordinates for fitting splines
+ * @param[in] y column_view of dependent variables to be fit along t axis
+ * @param[in] ids of incoming coordinate sets
+ * @param[in] offsets the exclusive scan of the spline sizes, prefixed by
+ * 0. For example, for 3 splines of 5 vertices each, the offsets input array
+ * is {0, 5, 10, 15}.
+ *
+ * @return cudf::table_view of coefficients for spline interpolation. The size
+ * of the table is ((M-n), 4) where M is `t.size()` and and n is 
+ * `ids.size()-1`.
+**/
+std::unique_ptr<cudf::experimental::table> cubicspline_coefficients(
+                                         cudf::column_view const& t,
+                                         cudf::column_view const& y,
+                                         cudf::column_view const& ids,
+                                         cudf::column_view const& offsets,
+                                         rmm::mr::device_memory_resource *mr,
+                                         cudaStream_t stream);
+
+std::unique_ptr<cudf::experimental::table> cubicspline_coefficients_default(
```

Comment:
What are these "default" functions?

Reply:
The default functions do not expose `stream` and `mr` - this is because the python layer doesn't have the ability to pass those args down, and can't overload function names, so they have to be uniquely named. I went through our existing CUDF Cython and couldn't find an alternative method.

Reply:
I've updated them to have the same name, but different namespaces! Problem solved.